File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
'''
특정 번호 전체가 다른 번호의 시작에 포함되어 있으면 접두어임 -> false
하나하나 비교 => n^2 시간초과

길이가 작은 것 부터 set에 넣기
포함 여부만 판단 해서 안에 있으면 false
끝까지 없으면 true

'''
def solution(phone_book):
    p_set = set()
    # 정렬
    phone_book.sort(key=lambda x: len(x))
    #print(phone_book)
    for i, s in enumerate(phone_book):
        for j in range(len(s)):
            #print(s[:j+1])
            if s[:j+1] in p_set:
                return False
        p_set.add(s)
        #print(p_set)
    return True

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
class Solution {
    public boolean solution(String[] phone_book) {
        /*
        오름차순 정렬 후
        
        각 번호를 
        시작점은 고정하고 길이별로 나눈 것을 set에 비교하기 
        있으면 바로 false임
        
        마지막에 모두 없으면 본인도 다시 넣기
        */
        Arrays.sort(phone_book, Comparator.comparingInt(String::length));
        Set<String> pre = new HashSet();
        for(String phone_num: phone_book){
            //System.out.println(phone_num);
            for(int i=1; i<=phone_num.length(); i++){
                String sub = phone_num.substring(0,i);
                if(pre.contains(sub)){
                    return false;
                }
            }
            pre.add(phone_num);
            //System.out.println(pre);
        }
        return true;
    }
}